# Comparisons

In [85]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from seaborn import set_style
from datetime import datetime, timedelta
set_style("whitegrid")

In [125]:
AHH = pd.read_csv('data/AHH.csv')
CIB = pd.read_csv('data/CIB.csv')
MCI = pd.read_csv('data/MCI.csv')
ALOT = pd.read_csv('data/ALOT.csv')
EMKR = pd.read_csv('data/EMKR.csv')
HHC = pd.read_csv('data/HHC.csv')
NSC = pd.read_csv('data/NSC.csv')
UTL = pd.read_csv('data/UTL.csv')
sp500 = pd.read_csv('data/SP500.csv')

In [126]:
AHH=AHH[AHH['y']==1]
CIB=CIB[CIB['y']==1]
MCI=MCI[MCI['y']==1]
ALOT=ALOT[ALOT['y']==1]
EMKR=EMKR[EMKR['y']==1]
HHC=HHC[HHC['y']==1]
NSC=NSC[NSC['y']==1]
UTL=UTL[UTL['y']==1]
sp500=sp500[sp500['y']==1]

In [127]:
stocks = [AHH,CIB,MCI,ALOT,EMKR,HHC,NSC,UTL,sp500]

In [141]:
for stock in stocks:
    stock.columns = stock.columns.str.strip()

In [143]:
for stock in stocks:
    stock['diff_fcmo'] = (stock['Friday_Close']-stock['Monday_Open'])

In [151]:
for stock in stocks:
    stock['Friday_date'] = pd.to_datetime(stock['Friday_date'])

In [157]:
AHH=AHH[AHH['Friday_date'].dt.month>=7]
CIB=CIB[CIB['Friday_date'].dt.month>=7]
MCI=MCI[MCI['Friday_date'].dt.month>=7]
ALOT=ALOT[ALOT['Friday_date'].dt.month>=7]
EMKR=EMKR[EMKR['Friday_date'].dt.month>=7]
HHC=HHC[HHC['Friday_date'].dt.month>=7]
NSC=NSC[NSC['Friday_date'].dt.month>=7]
UTL=UTL[UTL['Friday_date'].dt.month>=7]
sp500=sp500[sp500['Friday_date'].dt.month>=7]


In [160]:
stock_names = ['AHH','CIB','MCI','ALOT','EMKR','HHC','NSC','UTL','sp500']
p_val = [('AHH',0.008972), ('CIB',0.041472), ('MCI',0.005252), ('ALOT',0.041177),('EMKR',0.000603),('HHC',0.043817), ('UTL',0.041104),('UTL',0.049185)]

In [27]:
sorted(p_val, key=lambda x: x[1])

[('EMKR', 0.000603),
 ('MCI', 0.005252),
 ('AHH', 0.008972),
 ('UTL', 0.041104),
 ('ALOT', 0.041177),
 ('CIB', 0.041472),
 ('HHC', 0.043817),
 ('UTL', 0.049185)]

In [55]:
sum_inverse = 0
for i in range(len(p_val)):
    print(1/p_val[i][1])
    sum_inverse+=1/p_val[i][1]

111.45786892554614
24.11265432098765
190.4036557501904
24.285402044830853
1658.374792703151
22.822192299792317
24.328532502919423
20.33140185015757


In [56]:
sum_inverse

2076.1165003975752

In [57]:
p_val_norm=[(name,value,(1/value)/sum_inverse) for name, value in p_val]


In [58]:
p_val_norm

[('AHH', 0.008972, 0.05368574880272953),
 ('CIB', 0.041472, 0.011614306960312723),
 ('MCI', 0.005252, 0.09171145054419066),
 ('ALOT', 0.041177, 0.011697514103943691),
 ('EMKR', 0.000603, 0.7987869622853886),
 ('HHC', 0.043817, 0.010992732004885987),
 ('UTL', 0.041104, 0.011718288688645613),
 ('UTL', 0.049185, 0.00979299660990321)]

In [61]:
df=pd.DataFrame(p_val_norm)

In [62]:
df

0         1         2
0   AHH  0.008972  0.053686
1   CIB  0.041472  0.011614
2   MCI  0.005252  0.091711
3  ALOT  0.041177  0.011698
4  EMKR  0.000603  0.798787
5   HHC  0.043817  0.010993
6   UTL  0.041104  0.011718
7   UTL  0.049185  0.009793

In [63]:
df.columns=['ticker','p_val','1/p_val%']

In [64]:
df

ticker     p_val  1/p_val%
0    AHH  0.008972  0.053686
1    CIB  0.041472  0.011614
2    MCI  0.005252  0.091711
3   ALOT  0.041177  0.011698
4   EMKR  0.000603  0.798787
5    HHC  0.043817  0.010993
6    UTL  0.041104  0.011718
7    UTL  0.049185  0.009793

In [72]:
sp500[['Friday_date','Friday_Close','Monday_Open','diff_fcmo']]

Friday_date  Friday_Close  Monday_Open  diff_fcmo
7    2019-03-08       2743.07      2747.61      -4.54
38   2019-11-01       3066.91      3078.96     -12.05
54   2020-03-06       2972.37      2863.89     108.48
84   2020-10-30       3269.96      3296.20     -26.24
99   2021-03-12       3943.34      3942.96       0.38
129  2021-11-05       4697.53      4701.48      -3.95
144  2022-03-11       4204.31      4202.75       1.56
173  2022-11-04       3770.55      3780.71     -10.16
187  2023-03-10       3861.59      3835.12      26.47
217  2023-11-03       4358.34      4364.27      -5.93
231  2024-03-08       5123.69      5111.96      11.73
262  2024-11-01       5728.80      5725.15       3.65

In [73]:
# Buying the stock at Friday close on 1 Nov, 2019 and just holding till 1 Nov, 2024, we have an
# 86.7% increase
(5725.15-3066.91)/3066.91

0.8667486166858499

In [177]:
AHH

Unnamed: 0 Friday_date Monday_date  Friday_Open  Monday_Open  \
38           38  2019-11-01  2019-11-04        18.80        18.73   
84           84  2020-10-30  2020-11-02         8.58         9.18   
129         129  2021-11-05  2021-11-08        15.14        15.64   
173         173  2022-11-04  2022-11-07        11.29        11.35   
217         217  2023-11-03  2023-11-06        10.40        10.50   
262         262  2024-11-01  2024-11-04        10.93        10.78   

     Friday_High  Monday_High  Friday_Low  Monday_Low  Friday_Close  \
38         18.89        18.77       18.63       18.45         18.77   
84          9.17         9.24        8.58        8.63          9.01   
129        15.63        15.64       15.08       15.19         15.55   
173        11.46        11.50       11.09       11.16         11.31   
217        10.65        10.57       10.40       10.35         10.52   
262        10.97        10.95       10.71       10.74         10.75   

     Monday_Close  Friday_Adj Close  Monday_Adj Close  Friday_Volume  \
38          18.50             14.09             13.89         174800   
84           8.85              7.10              6.97         692800   
129         15.29             12.82             12.61         318900   
173         11.40              9.84              9.92         293700   
217         10.44              9.80              9.72         390100   
262         10.91             10.75             10.91         485100   

     Monday_Volume  y  diff_fcmo  
38          239600  1       0.04  
84          354900  1      -0.17  
129         272500  1      -0.09  
173         303300  1      -0.04  
217         225000  1       0.02  
262         609300  1      -0.03

In [190]:
end=[]
for (stock,pval) in zip(stocks,p_val_norm):
    # Initial cash and shares
    cash = 100*pval[2]  # Start with $100 times the pval normalized percent
    shares = 0  # Initially no shares

    # Simulate the strategy
    for index, row in stock.iterrows():
        friday_price = row['Friday_Close']
        monday_price = row['Monday_Open']

        # Sell all shares on Friday
        cash += shares * friday_price
        shares = 0  # Reset shares to 0 after selling

        # Buy on Monday if the price is lower than Friday
        if monday_price < friday_price:
            shares = cash / monday_price  # Buy as many shares as possible
            cash -= shares * monday_price  # Update cash after buying

    # Final portfolio value
    total_value = cash + shares * stock.iloc[-1]['Friday_Close']

    # Calculate gain/loss
    initial_investment = 100*pval[2]
    gain_loss = total_value - initial_investment

    end.append((pval[0],total_value))

    print('Stock: '+ pval[0])
    print(f"Initial Investment: ${100*pval[2]:.2f}")
    print(f"Final Cash: ${cash:.2f}")
    print(f"Shares Owned: {shares}")
    print(f"Final Portfolio Value: ${total_value:.2f}")
    print(f"Gain/Loss: ${gain_loss:.2f}")
    print()


Stock: AHH
Initial Investment: $5.37
Final Cash: $2.81
Shares Owned: 0
Final Portfolio Value: $2.81
Gain/Loss: $-2.56

Stock: CIB
Initial Investment: $1.16
Final Cash: $0.64
Shares Owned: 0
Final Portfolio Value: $0.64
Gain/Loss: $-0.52

Stock: MCI
Initial Investment: $9.17
Final Cash: $0.00
Shares Owned: 0.3560727564764476
Final Portfolio Value: $7.09
Gain/Loss: $-2.08

Stock: ALOT
Initial Investment: $1.17
Final Cash: $0.00
Shares Owned: 0.04484386893338941
Final Portfolio Value: $0.55
Gain/Loss: $-0.62

Stock: EMKR
Initial Investment: $79.88
Final Cash: $0.00
Shares Owned: 18.457820243616606
Final Portfolio Value: $36.92
Gain/Loss: $-42.96

Stock: HHC
Initial Investment: $1.10
Final Cash: $0.00
Shares Owned: 0.009294039769904354
Final Portfolio Value: $0.70
Gain/Loss: $-0.40

Stock: UTL
Initial Investment: $1.17
Final Cash: $0.00
Shares Owned: 0.008618373443277245
Final Portfolio Value: $2.16
Gain/Loss: $0.98

Stock: UTL
Initial Investment: $0.98
Final Cash: $0.00
Shares Owned: 0.00

In [191]:
end

[('AHH', 2.811559144341894),
 ('CIB', 0.6396754106902709),
 ('MCI', 7.0894085814460714),
 ('ALOT', 0.5506827105020219),
 ('EMKR', 36.91564048723321),
 ('HHC', 0.7017000026277789),
 ('UTL', 2.155369014429206),
 ('UTL', 0.5565129941886356)]

In [192]:
end_sum=0
for price in end:
    end_sum+=price[1]
end_sum

51.420548345459096

If we start with $100 and put in amounts to the 8 stocks proportional to the inverse p-values, we can follow a strategy of selling shares on Friday at closing, then buying on Monday if it's favorable. Following this strategy, we end up with $51.42 in our portfolio.

But just holding $100 in the S&P500, we would have $186.67. It seems that these particular DST weekends are seeing a reverse weekend effect.

In [193]:
sp500

Unnamed: 0 Friday_date Monday_date  Friday_Open  Monday_Open  \
38           38  2019-11-01  2019-11-04      3050.72      3078.96   
84           84  2020-10-30  2020-11-02      3293.59      3296.20   
129         129  2021-11-05  2021-11-08      4699.26      4701.48   
173         173  2022-11-04  2022-11-07      3766.98      3780.71   
217         217  2023-11-03  2023-11-06      4334.23      4364.27   
262         262  2024-11-01  2024-11-04      5723.22      5725.15   

     Friday_High  Monday_High  Friday_Low  Monday_Low  Friday_Close  \
38       3066.95      3085.20     3050.72     3074.87       3066.91   
84       3304.93      3330.14     3233.94     3279.74       3269.96   
129      4718.50      4714.92     4681.32     4694.39       4697.53   
173      3796.34      3813.95     3708.84     3764.70       3770.55   
217      4373.62      4372.21     4334.23     4347.53       4358.34   
262      5772.52      5741.43     5723.22     5696.51       5728.80   

     Monday_Close  Friday_Adj Close  Monday_Adj Close  Friday_Volume  \
38        3078.27           3066.91           3078.27     3956290000   
84        3310.24           3269.96           3310.24     4843930000   
129       4701.70           4697.53           4701.70     4467180000   
173       3806.80           3770.55           3806.80     5400180000   
217       4365.98           4358.34           4365.98     4570960000   
262       5712.69           5728.80           5712.69     3796450000   

     Monday_Volume  y  diff_fcmo  
38      4178040000  1     -12.05  
84      4314690000  1     -26.24  
129     4269710000  1      -3.95  
173     4341620000  1     -10.16  
217     3656340000  1      -5.93  
262     3602060000  1       3.65